lib/http2.c

```diff
@@ -1999,26 +2000,33 @@ static ssize_t cf_h2_send(struct Curl_cfilter *cf, struct Curl_easy *data,
     rv = nghttp2_session_resume_data(ctx->h2, stream->stream_id);
     if(nghttp2_is_fatal(rv)) {
       *err = CURLE_SEND_ERROR;
-      len = -1;
+      nwritten = -1;
       goto out;
     }
     result = h2_session_send(cf, data);
     if(result) {
       *err = result;
-      len = -1;
+      nwritten = -1;
       goto out;
     }
-    len -= stream->upload_len;
 
-    /* Nullify here because we call nghttp2_session_send() and they
-       might refer to the old buffer. */
+    /* The callback for stream input updates `upload_len` with what it
+     * has provided to nghttp2. If it has *not* been called, `upload_len`
+     * is unchanged and we should report a block on send. */
+    if(stream->upload_len == len) {
+      *err = CURLE_AGAIN;
+      nwritten = -1;
+      goto out;
+    }
+
+    nwritten = len - stream->upload_len;
```

Comment:
I get no warning for this despite the unsigned/signed conversion

Reply:
the result of a `-` operation is signed.

Reply:
? You are probably thinking of some other type or language. The result should be unsigned which is then assigned to signed type. VS of any version doesn't seem to warn on this at all (I've asked [here](https://stackoverflow.com/questions/75445208/can-visual-studio-warn-of-unsigned-signed-type-conversion) anyway) and gcc only warns when `-Wsign-conversion`, so it looks like a non issue.

Reply:
You are right. I am tempted to edit my stupid reply above, but will accept my shortcomings.